### PR TITLE
Game widget

### PR DESCRIPTION
**Pull request by renancaraujo**
# Summary

This is a major (and very breaking) change.

**This problem:** our current widgets overlay API uses the anti-pattern of storing widget instances into the state. This can be a problem for stateful overlays that are added and removed from the tree several times. Also, there is no way to provide extra properties to a game that could only be defined via the build context (like text direction). 

**The approach:** we are changing from `game.widget` approach into something like this:
```dart
GameWidget(
  game: game
)
```

This enabled  some changes in some existing features and also the creation of new ones:

### Widgets overlay API

The one that was the one that has changed the most. Overlay widgets are now declared in the widget tree rather than the in-game methods like before.

**Before**:
Overlays were declared alongside its widgets via an API that didnt allow for context access and the instances here delcared were saved internally to a widget state, making it possibly "dirty" of being rendered again:
```dart
addWidgetOverlay(
  "PauseMenu", // Your overlay identifier
  Center(child:
      Container(
          width: 100,
          height: 100,
          color: const Color(0xFFFF0000),
          child: const Center(child: const Text("Paused")),
      ),
  ) // Your widget, this can be any Flutter widget
);

removeWidgetOverlay("PauseMenu"); // Use the overlay identifier to remove the overlay
```

**After**:
Overlays are declared alongside its widgets, and their visibility is still controlled via in-game methods, like before:

```dart
// inside game methods:

final pauseOverlayIdentifier = "PauseMenu";
overlays.add(pauseOverlayIdentifier); // marks "PauseMenu" to be renderable.
overlays.remove(pauseOverlayIdentifier); // marks "PauseMenu" to not be renderable.

// on the widget declaration

return GameWidget.withOverlay(
  game: MyGameSubClass() // MyGameSubClass has to be mixed with `HasWidgetsOverlay` 
  overlayBuilderMap: {
    "PauseMenu": (context) { // notice the overlay identifier
      return Positioned.fill( // Overlays are rendered directly under a stack, so Positioned are eligible
        child: Text("This is a pause menu")
      );
    },
  },
)
```

### Loading widget

`Game` has an `onLoad` method that is used to return a `Future` for async game asset loading.

**Before**:
`Game` had an overridable property `.loadingWidget` that returned a Widget to be shown in the place of the game whilst that `Future` was not resolved.
```dart
// game.dart
Widget loadingWidget() => Container();
```
**After:**
The loading widget has been replaced by `loadingBuilder` on the `GameWidget` declaration. This allows for context access. Also, an argument "error" was added to make it possible to show a widget for the scenario where the `Future` goes south:
```dart
GameWidget(
  game: game,
  loadingBuilder: (ctx, hasError) {
    return hasError ? Text("oh it is an error") : Text("oh it is loading")
  },
)
```

### Game attachment verification

Previously, our API didn't stop the engine user to attach the same game instance more than once into the widget tree. This would be a problem since for every attachment, we have an instance of `GameLoop` and it would result in **several updates and paints** for each frame, which is extremely expensive.
Now we verify if a game instance had been attached before adding it into a widget tree. We do verify by saving the build context into the game instance, which leads us to the next point.

### Build context on game

We make the build context to be accessible inside a game instance via the `buildContext` property. This enables access to contextual stuff inside out game/component system. This is important for gamification use cases where the game state can use (or even be) subsets of the application state. 

### Background builder

Like overlay, but with no name, and it is behind the game elements, not above.

# Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [ ] This branch is based on the latest `master`
- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [x] I have made corresponding changes to the documentation
- [x] I have added examples for new features in `doc/examples`
- [ ] The continuous integration (CI) is passing

### CODE REVIEW

File: lib/game/game.dart
Comment (spydon) [inline lib/game/game.dart:118]: What happens here if you have for example a dynamic amount of widgets, do you have to assign a unique string to each one of them then? There must be a better way to do this than to use strings?

File: doc/game.md
Comment (erickzanardo) [inline doc/game.md:11]: Top instead of topo

Comment (erickzanardo) [inline doc/game.md:11]: Widget structure is repeated here

Comment (erickzanardo) [inline doc/game.md:11]: Etc instead of whatnot

File: doc/game.md
Comment (erickzanardo) [inline doc/game.md:27]: This "for advanced the Forge2d" is sounding strange, not sure what you meant here

File: doc/game.md
Comment (erickzanardo) [inline doc/game.md:71]: Rendered instead of renderable

File: lib/game/game.dart
Comment (erickzanardo) [inline lib/game/game.dart:116]: Since this is always available on the gamewidget, I would take a leap here and say that we don't need this anymore, maybe we could just make this a default feature on the game class.

File: lib/game/game.dart
Comment (erickzanardo) [inline lib/game/game.dart:58]: Why not just use onAttach?

File: lib/game/game.dart
Comment (erickzanardo) [inline lib/game/game.dart:77]: Why not just use onDetach?

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:58]: Attach is the action. Prefixing with `on` normally refers to the reactions, which are triggered internally.

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:77]: Ditto.

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:58]: Mind that onAttach still exists.

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:118]: We have to apply that dynamic in a widget build context.  In order to have an active control of which overlay is active on the game side, we have to identify the overlays. 

>There must be a better way to do this than to use strings?
What is the problem with strings?

File: lib/game/game.dart
Comment (erickzanardo) [reply lib/game/game.dart:58]: I am not sure with both needs to exists, `attach` calls `onAttach`, I think since `onAttach` has `@mustCallsuper` it is fine to be with just that method.

But I have no strong opinion on this, so I leave it up to you

Comment (renancaraujo) [reply lib/game/game.dart:58]: An API with only `onAttach` would be too much error-prone. What if the user does stuff before calling super? What is the point of passing the build context and pipeline owner to a user-provided callback?

Comment (erickzanardo) [reply lib/game/game.dart:58]: Like I said, it is up to you 😉

File: lib/game/game.dart
Comment (erickzanardo) [reply lib/game/game.dart:118]: I also see no problems with strings for keys

File: doc/examples/animations/lib/main.dart
Comment (luanpotter) [inline doc/examples/animations/lib/main.dart:18]: idea: how about we provide a runGame helper function that just calls runApp with an empty GameWidget? that would make most of our examples a bit simpler. of course if you want to customize anything on the widget then you do runApp with the game widget

File: doc/examples/effects/infinite_effects/lib/main.dart
Comment (luanpotter) [inline doc/examples/effects/infinite_effects/lib/main.dart:70]: this change is wrong

File: doc/examples/with_widgets_overlay/lib/example_game.dart
Comment (luanpotter) [inline doc/examples/with_widgets_overlay/lib/example_game.dart:24]: how about show and hide? it feels more close to the behavior than add/remove. but I am ok with either

File: doc/game.md
Comment (luanpotter) [inline doc/game.md:8]:
```suggestion
The class `Game` can be subclassed and will provide, among others, these gameloop methods.
```

File: doc/game.md
Comment (luanpotter) [inline doc/game.md:9]: `At the same time` is odd here, at the same time as what?

File: doc/game.md
Comment (luanpotter) [inline doc/game.md:58]: can be wrapped in a widget?

Comment (luanpotter) [reply doc/game.md:58]: `But there is a the Widgets overlay API` doesn't make sense to me

File: doc/game.md
Comment (luanpotter) [inline doc/game.md:60]: to be shown

File: lib/game/embedded_game_widget.dart
Comment (luanpotter) [inline lib/game/embedded_game_widget.dart:8]: is this class still needed? GameWidget will work on all cases now right?

File: lib/game/game.dart
Comment (luanpotter) [reply lib/game/game.dart:118]: strings are not type safe. but I think it would be overkill to do something else. like a typed enum, and dart enums aren't as powerful as java's to make this sort of thing easy. I am ok with the string as well

Comment (luanpotter) [reply lib/game/game.dart:118]: we do have to throw a runtime error if it doesn't exist do the user can fix it.

File: lib/game/game_widget.dart
Comment (luanpotter) [inline lib/game/game_widget.dart:33]: `o use it see`?

File: lib/game/game_widget.dart
Comment (luanpotter) [inline lib/game/game_widget.dart:86]: why not an optional this.overlayBuilderMap, instead of two constructors?

**Review by luanpotter — COMMENTED** [review]
Added some comments and questions

File: doc/examples/with_widgets_overlay/lib/example_game.dart
Comment (renancaraujo) [reply doc/examples/with_widgets_overlay/lib/example_game.dart:24]: Show and hide would give the wrong idea that only the visibility of the overlays would change when they are actually completely removed from the widget tree.

File: doc/examples/effects/infinite_effects/lib/main.dart
Comment (renancaraujo) [reply doc/examples/effects/infinite_effects/lib/main.dart:70]: Totally
disagree.

File: doc/examples/animations/lib/main.dart
Comment (renancaraujo) [reply doc/examples/animations/lib/main.dart:18]: Nice idea, I wouldn't add this here since this can scale for more stuff on the bootstrap of the app.

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:118]: This can be solved with generics, but generics in dart are like all my cousins: ugly.

File: lib/game/game_widget.dart
Comment (renancaraujo) [reply lib/game/game_widget.dart:86]: When using overlay, we assert if the game has the mixin.

File: lib/game/game_widget.dart
Comment (renancaraujo) [reply lib/game/game_widget.dart:33]: Hardcore typo there.

File: lib/game/embedded_game_widget.dart
Comment (renancaraujo) [reply lib/game/embedded_game_widget.dart:8]: pretty much yes. This is our render box widget, for the overlay stuff we need a stateful one, they have very different concerns.

File: doc/game.md
Comment (renancaraujo) [reply doc/game.md:9]: Too many Linus tech tips videos. I'm talking just like him.

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:118]: Added runtime verification if the overlay exists.

File: lib/game/game_widget.dart
Comment (luanpotter) [reply lib/game/game_widget.dart:86]: good point

File: lib/game/embedded_game_widget.dart
Comment (luanpotter) [reply lib/game/embedded_game_widget.dart:8]: I thought everyone was going to use `GameWidget(game: game)` now instead of `EmbeddedGameWidget(game.widget)`. This was used to add the game as a widget inside a regular app. The current `GameWidget` already provide that right?

**Review by luanpotter — APPROVED** [review]
LGTM with one final question about Embeddable Widget

File: lib/game/game.dart
Comment (spydon) [inline lib/game/game.dart:125]: If the same one is added two times, shouldn't this return false?

File: doc/game.md
Comment (spydon) [reply doc/game.md:62]: "by doing so," shouldn't have a comma in the end
a `.overlays` -> the `.overlays`

File: doc/game.md
Comment (spydon) [inline doc/game.md:63]: First occurrence: a overlay -> an overlay
Second occurrence: a overlay -> the overlay

File: doc/game.md
Comment (spydon) [inline doc/game.md:64]: represents -> represent
Both commas should be removed,

File: lib/game/game_widget.dart
Comment (spydon) [inline lib/game/game_widget.dart:16]: Should be T instead of G

**Review by spydon — COMMENTED** [review]
Looks nice! 
I don't like the String set, but I don't have a cleaner solution as a suggestion either.

File: lib/game/game.dart
Comment (spydon) [reply lib/game/game.dart:118]: Just curious, how would you solve this with generics?

File: lib/game/game_widget.dart
Comment (renancaraujo) [reply lib/game/game_widget.dart:16]: Why?

File: lib/game/embedded_game_widget.dart
Comment (renancaraujo) [reply lib/game/embedded_game_widget.dart:8]: EmbeddedGameWidget is internal and should not be used outside the package.

File: lib/game/game_widget.dart
Comment (spydon) [reply lib/game/game_widget.dart:16]: It is the standard to denote type with `T`, I haven't seen that it would be any differently in Dart yet at least.
They use `T` here too: https://dart.dev/guides/language/effective-dart/style
It helps to know that the type is derived for generics and that it isn't the name of an existing type when reading the code.

Comment (renancaraujo) [reply lib/game/game_widget.dart:16]: > use T, S, and U for generics that have a single type parameter and where the surrounding type makes its meaning obvious

`GameWidget` makes it pretty easy to understand that it is about a `Game`. Wilco thanx!

File: lib/game/game.dart
Comment (renancaraujo) [reply lib/game/game.dart:118]: The user would pass an enum via a generic and we would define this generic as the key type on our overlay builder map. 
But this would require a generic on the game class as well. Also, would be quite weird since we cant [constrain a generic to be an enum](https://stackoverflow.com/questions/59362403/dart-constrain-a-generic-type-t-to-an-enum).

File: lib/game/embedded_game_widget.dart
Comment (renancaraujo) [reply lib/game/embedded_game_widget.dart:8]: Removed it in favor of a private class `_GameRenderObjectWidget`.

**Review by spydon — APPROVED** [review]
Just some final grammar comments!

**Review by erickzanardo — APPROVED** [review]
LGTM, can't wait for this to be merged

**Pull request merged**
